### PR TITLE
IM5K4 PMPS with SP1K4 FZP and SP1K4 ATT removing two targets

### PR DESCRIPTION
**Pull request by tongju12**
## Description

Add PMPS into IM5K4 with Sp1K4 FZP veto and remove target 4,5 of Sp1K4-ATT

IM5K4: adding IM5K4 is vetoed by SP1K4 FZP all targets
Sp1K4 ATT, remove target 4 and 5 by scientists request because target4 and 5 bumping inside the attenuator
Change Sp1K4 FZP Y negative limit switch

## Motivation and Context

Scientists request these

## How Has This Been Tested?

Yes, during last night beam time, all tested. 

## Where Has This Been Documented?

(https://jira.slac.stanford.edu/browse/ECS-5489)

## Pre-merge checklist
- [ ] Code works interactively
- [ ] Code contains descriptive comments
- [ ] Test suite passes locally
- [ ] Libraries are set to fixed versions and not ``Always Newest``
- [ ] Code committed with ``pre-commit`` (alternatively ``pre-commit run --all-files``)

### CODE REVIEW

**Comment by ZLLentz** [comment]
Hi Tong, can you please fill out the rest of the PR description template? I'll give this a review once you finish filling it out.
I'm also going to edit back in the deleted "description" header so I don't get confused when I look at this again.

**Comment by tongju12** [comment]
Zach,
I just finish all the other fields. Sorry, I thought I already did.

**Comment by ZLLentz** [comment]
Hey Tong, it's a smaller detail but you should also stop deleting the `## Description` text from the content. This is what creates the nice section headers. I'll put it back for you.

File: plc-tmo-motion/_Config/NC/Axes/ZonePlateY.xti [hunk 1]
```diff
@@ -1324,7 +1324,7 @@ External Setpoint Generation:
 		</AxisPara>
 		<Encoder Name="Enc" EncType="29">
 			<EncPara ScaleFactorNumerator="5e-05" Offset="-214733.3875" MaxCount="#xffffffff">
-				<SoftEndMinControl Enable="true" Range="-9.6"/>
+				<SoftEndMinControl Enable="true" Range="-10.6"/>
```
Comment (ZLLentz) [inline plc-tmo-motion/_Config/NC/Axes/ZonePlateY.xti:1327]: Undocumented change: Y soft limit moved (unless this is what you meant by the limit switch change?)

File: plc-tmo-motion/tmo_motion/POUs/PRG_IM5K4_PPM.TcPOU [hunk 2]
```diff
@@ -38,34 +38,7 @@ fbStateSetup(StPositionState:=fbIM5K4.stPower, fposition:=-44.23, sPmpsState :=
...
-fbIM5K4.stYag1.bValid := TRUE;
-
-
-//fbIM5K4.stYag2.fPosition := -94.24;
-fbIM5K4.stYag2.fVelocity := fStartupVelo;
-fbIM5K4.stYag2.bUseRawCounts := FALSE;
-fbIM5K4.stYag2.bMoveOk := TRUE;
-fbIM5K4.stYag2.bValid := TRUE;
-
-*}
```
Comment (ZLLentz) [inline plc-tmo-motion/tmo_motion/POUs/PRG_IM5K4_PPM.TcPOU:68]: Undocumented good maintenance fix: removed various commented out old code

File: plc-tmo-motion/tmo_motion/POUs/PRG_LI2K4_IP1.TcPOU [hunk 3]
```diff
@@ -54,8 +54,8 @@ fbMotionLI2K4X(stMotionStage:=Main.M46);
 //LI2K4 2DPMPS
 fbStateSetup(stPositionState:=stDefault, bSetDefault:=TRUE);
 fbStateSetup(stPositionState:=aLI2K4XStates[ENUM_LaserCoupling_States.OUT], fDelta := 0.5, sName := 'OUT', fPosition:=-0.75);
-
-{*aLI2K4XStates[ENUM_LaserCoupling_States.OUT].nMoveOrder := 1;
+{*
+aLI2K4XStates[ENUM_LaserCoupling_States.OUT].nMoveOrder := 1;
```
Comment (ZLLentz) [inline plc-tmo-motion/tmo_motion/POUs/PRG_LI2K4_IP1.TcPOU:58]: Undocumented change, I'm not sure I understand what this one is. (I could guess what the others were for)

File: plc-tmo-motion/tmo_motion/POUs/PRG_IM5K4_PPM.TcPOU [hunk 4]
```diff
@@ -76,15 +49,18 @@ ELSE
...
-   fbIM5K4.stOut.fPosition := -5.13;
-   fbIM5K4.stPower.fPosition := -44.23;
-   fbIM5K4.stYag1.fPosition := -68.23;
-   fbIM5K4.stYag2.fPosition := -94.24;
-END_IF
-*}
+//IM5K4 is vetoed by SP1K4-FZP
+CASE GVL_TcGVL.eSP1K4FZP OF
+    ENUM_ZonePlate_States.FZP290_1, ENUM_ZonePlate_States.FZP290_2, ENUM_ZonePlate_States.FZP410_1, ENUM_ZonePlate_States.FZP410_2, ENUM_ZonePlate_States.FZP460_1, ENUM_ZonePlate_States.FZP460_2, ENUM_ZonePlate_States.FZP530_1,
+    ENUM_ZonePlate_States.FZP530_2, ENUM_ZonePlate_States.FZP750_1, ENUM_ZonePlate_States.FZP750_2, ENUM_ZonePlate_States.FZP860_1, ENUM_ZonePlate_States.FZP860_2, ENUM_ZonePlate_States.FZP860_3, ENUM_ZonePlate_States.Yag:
```
Comment (ZLLentz) [inline plc-tmo-motion/tmo_motion/POUs/PRG_IM5K4_PPM.TcPOU:55]: I double-checked and this is definitely all of the "in" states :+1:
It's a bit wordy but this is still probably the correct way to implement this to guard against invalid states

File: plc-tmo-motion/tmo_motion/POUs/PRG_SP1K4.TcPOU [hunk 5]
```diff
@@ -86,7 +86,7 @@ VAR
     att_enumGet: ENUM_SolidAttenuator_States;
     fbATTSetup: FB_StateSetupHelper;
     fbATTDefault: ST_PositionState := (
-        fDelta:=0.5,
+        fDelta:=1.5,
```
Comment (ZLLentz) [inline plc-tmo-motion/tmo_motion/POUs/PRG_SP1K4.TcPOU:89]: Undocumented change again, but I think it's clear why you'd want this

File: plc-tmo-motion/tmo_motion/tmo_motion.plcproj [hunk 6]
```diff
@@ -216,7 +216,7 @@
       <Resolution>lcls-twincat-common-components, 3.4.0 (SLAC)</Resolution>
     </PlaceholderResolution>
     <PlaceholderResolution Include="lcls-twincat-motion">
-      <Resolution>lcls-twincat-motion, 4.0.9 (SLAC)</Resolution>
+      <Resolution>lcls-twincat-motion, 4.1.1 (SLAC)</Resolution>
```
Comment (ZLLentz) [inline plc-tmo-motion/tmo_motion/tmo_motion.plcproj:219]: Undocumented change again, another good one

File: plc-tmo-motion/tmo_motion/tmo_motion.plcproj
Comment (ZLLentz) [inline plc-tmo-motion/tmo_motion/tmo_motion.plcproj:1]: Not this file, but pre-commit is failing on something really minor from a previous PR. I'm ok with ignoring it.

**Review by ZLLentz — APPROVED** [review]
Hi Tong, this all looks good, but you made a lot of changes that aren't in the PR description. It's a good idea to document that these were done in this PR (and why). The code itself looks good to me.
